plutus-contract/src/Language/Plutus/Contract/Trace.hs

```diff
@@ -526,29 +542,88 @@ notifySlot
-            Just slot | slot <= currentSlot -> Just (AwaitSlot.event currentSlot)
-            _                               -> Nothing
+    currentSlot <- getCurrentSlot wallet
+    let handleEvent e = case AwaitSlot.request e of
+            Just waitingSlot | currentSlot >= waitingSlot -> Just (AwaitSlot.event currentSlot)
+            _                                             -> Nothing
     respondToRequest wallet handleEvent
 
+getCurrentSlot :: (MonadState EmulatorState m) => Wallet -> ContractTrace s e m a Slot
+getCurrentSlot wallet = lift $ gets (view (EM.walletClientStates . at wallet . non EM.emptyNodeClientState . EM.clientSlot))
+
+-- | Whether to update the internal slot counter of a wallet (that is, the wallet's notion
+--   of the current time)
+data SlotNotifications =
+        SendSlotNotifications
+        | DontSendSlotNotifications
+    deriving (Eq, Ord, Show)
+
+-- | @defaultSlotNotifications = SendSlotNotifications@
+defaultSlotNotifications :: SlotNotifications
+defaultSlotNotifications = SendSlotNotifications
+
 -- | Add a number of empty blocks to the blockchain.
+addBlocks'
+    :: ( MonadEmulator (TraceError e) m
+       , AwaitSlot.HasAwaitSlot s
+       )
+    => SlotNotifications -- ^ Whether to notify all wallets after each block.
+    -> Integer -- ^ How many blocks to add.
+    -> ContractTrace s e m a ()
+addBlocks' n i = do
+    let notify = case n of
+            SendSlotNotifications     -> traverse_ notifySlot allWallets
+            DontSendSlotNotifications -> pure ()
+
+    replicateM_ (fromIntegral i) $ do
+        void $ lift $ EM.processEmulated (EM.addBlocksAndNotify allWallets 1)
+        notify
+
+-- | Add a number of empty blocks to the blockchain, updating each wallet's
+--   slot after each block
 addBlocks
-    :: ( MonadEmulator (TraceError e) m )
-    => Integer
+    :: ( MonadEmulator (TraceError e) m
+       , AwaitSlot.HasAwaitSlot s
+       )
+    => Integer -- ^ How many blocks to add.
     -> ContractTrace s e m a ()
-addBlocks i =
-    void $ lift $ EM.processEmulated (EM.addBlocksAndNotify allWallets i)
+addBlocks = addBlocks' SendSlotNotifications
 
 -- | Add blocks until the given slot has been reached.
 addBlocksUntil
-    :: ( MonadEmulator (TraceError e) m )
+    :: ( MonadEmulator (TraceError e) m
+       , AwaitSlot.HasAwaitSlot s
+       )
     => Slot
     -> ContractTrace s e m a ()
-addBlocksUntil sl = do
+addBlocksUntil = addBlocksUntil' SendSlotNotifications
+
+-- | Add blocks until the given slot has been reached.
+addBlocksUntil'
+    :: ( MonadEmulator (TraceError e) m
+       , AwaitSlot.HasAwaitSlot s
+       )
+    => SlotNotifications
+    -> Slot
+    -> ContractTrace s e m a ()
+addBlocksUntil' n sl = do
     currentSlot <- lift $ use (EM.chainState . EM.currentSlot)
     let Slot missing = sl - currentSlot
-    addBlocks (max 0 missing)
+    addBlocks' n (max 0 missing)
+
+-- | Options for 'handleBlockchainEvents'
+data HandleBlockchainEventsOptions =
```

Comment:
I think this is fine: at the moment we want this "handle blockchain events" function that just sweeps everything under the rug. In future we might want to revisit it but I think it's fine to offer a few knobs on the rug-sweeping.